### PR TITLE
Final Submission

### PR DESCRIPTION
**Pull request by ruil1**
Name: Rui Lin
SMU ID: 44367652

### CODE REVIEW

**Comment by ruil1** [comment]
Arrows do not work, you need to input letters instead of arrows to play the game.

Left: a
Right: d
Up: w
Down: s

File: terminal2048/test/terminal2048_test.exs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    test "has a reasonable status", c do
+      assert c.game.game_state    == :initializing
+      assert length(c.game.board) == 16
+      assert c.game.score         == 0
+    end
+  end
+
+  describe "move left" do
+    setup do
+      board = [ nil,  2,  nil,  2,
```
Comment (pragdave) [inline terminal2048/test/terminal2048_test.exs:20]: I like the way you laid this out.

# +2

File: terminal2048/lib/terminal2048/move.ex [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+  defp handle_game_state(true, _),      do: :lost
+  defp handle_game_state(_, true),      do: :won
+  defp handle_game_state(false, false), do: :continue
+
+  defp move(game, :left),  do: move_left({game.board, game.score})
+  defp move(game, :right), do: move_right({game.board, game.score})
+  defp move(game, :up),    do: move_up({game.board, game.score})
+  defp move(game, :down),  do: move_down({game.board, game.score})
+
```
Comment (pragdave) [inline terminal2048/lib/terminal2048/move.ex:51]: This is clever, but I think you could have refactored this to remove the duplicate code. For example, the first two could be:

~~~ elixir
  defp move_left_right({board, score}, maybe_reverse) do
    row_with_score = board
      |> Enum.chunk_every(4)
      |> Enum.map(maybe_reverse)
      |> Enum.map(&leftward_operation(&1))

    new_board = row_with_score
      |> Enum.map(&List.first/1)
      |> Enum.map(maybe_reverse)
      |> List.flatten

    new_score = row_with_score
      |> Enum.map(&List.last/1)
      |> List.flatten
      |> Enum.sum

    {new_board, score + new_score}
  end


  defp move_left({board, score}) do
    move_left_right({board, score}, &no_reverse/1)
  end

  defp move_right({board, score}) do
   move_left_right({board, score}, &Enum.reverse/1)
  end
  
  defp no_reverse(x), do: x
~~~

# -2

File: terminal2048/lib/terminal2048/player.ex [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  end
+
+  defp clear_screen(), do: IO.write("\e[H\e[2J")
+
+  defp get_move("w"), do: :up
+  defp get_move("s"), do: :down
+  defp get_move("a"), do: :left
+  defp get_move("d"), do: :right
+  defp get_move(_) do
+    "\nInvalid move. Try another move: "
```
Comment (pragdave) [inline terminal2048/lib/terminal2048/player.ex:45]: I don't like the duplication here.

If `display_board` was moved outside `get_next_move`, then you could simply call it from here. 

This is a good example of how a "code smell" (the duplication) is a pointer to a better design.

# -2

**Comment by pragdave** [comment]
I love playing 2048, so this way fun to grade.

I got a little confused, though. You have a Game module, and the Player code has lots of references to a `game` variable, but it is actually a reference to a `Board`. I tried deleting the entire Game module, and removing it from the supervisor, and the app still ran and passed its tests.

Unfortunately this means that you don't have a supervisor or GenServer in your code, which cost 10%.

You also lost just a few marks for some instances of duplication. This is something to think about going forward: whenever you find yourself writing the same thing twice, stop and think about the design. Normally you'll find the duplication is a symptom of something can can be fixed with a little refactoring.

This was a nice, tidy, project. Thank you

**Pull request closed without merging**